pypsa/optimization/abstract.py

```diff
@@ -0,0 +1,254 @@
+            np.sqrt((s_nom_prev - n.lines.s_nom_opt).pow(2).mean())
+            / n.lines["s_nom_opt"].mean()
+        )
+        logger.info(
+            f"Mean square difference after iteration {iteration} is " f"{lines_err}"
+        )
+        return lines_err
+
+    def save_optimal_capacities(n, iteration, status):
+        for c, attr in pd.Series(nominal_attrs)[n.branch_components].items():
+            n.df(c)[f"{attr}_opt_{iteration}"] = n.df(c)[f"{attr}_opt"]
+        setattr(n, f"status_{iteration}", status)
+        setattr(n, f"objective_{iteration}", n.objective)
+        n.iteration = iteration
+        n.global_constraints = n.global_constraints.rename(
+            columns={"mu": f"mu_{iteration}"}
+        )
+
+    if track_iterations:
+        for c, attr in pd.Series(nominal_attrs)[n.branch_components].items():
+            n.df(c)[f"{attr}_opt_0"] = n.df(c)[f"{attr}"]
+
+    iteration = 1
+    kwargs["store_basis"] = True
+    diff = msq_threshold
+    while diff >= msq_threshold or iteration < min_iterations:
+        if iteration > max_iterations:
+            logger.info(
+                f"Iteration {iteration} beyond max_iterations "
+                f"{max_iterations}. Stopping ..."
+            )
+            break
+
+        s_nom_prev = n.lines.s_nom_opt.copy() if iteration else n.lines.s_nom.copy()
+        status, termination_condition = n.optimize(snapshots, **kwargs)
+        assert status == "ok", (
+            f"Optimization failed with status {status}"
+            f"and termination {termination_condition}"
+        )
+        if track_iterations:
+            save_optimal_capacities(n, iteration, status)
+        update_line_params(n, s_nom_prev)
+        diff = msq_diff(n, s_nom_prev)
+        iteration += 1
+
+    logger.info("Running last lopf with fixed branches (HVDC links and HVAC lines)")
+
+    ext_dc_links_b = n.links.p_nom_extendable & (n.links.carrier == "DC")
+    s_nom_orig = n.lines.s_nom.copy()
+    p_nom_orig = n.links.p_nom.copy()
+
+    n.lines.loc[ext_i, "s_nom"] = n.lines.loc[ext_i, "s_nom_opt"]
+    n.lines.loc[ext_i, "s_nom_extendable"] = False
+
+    n.links.loc[ext_dc_links_b, "p_nom"] = n.links.loc[ext_dc_links_b, "p_nom_opt"]
+    n.links.loc[ext_dc_links_b, "p_nom_extendable"] = False
+
+    n.optimize(snapshots, **kwargs)
+
+    n.lines.loc[ext_i, "s_nom"] = s_nom_orig.loc[ext_i]
+    n.lines.loc[ext_i, "s_nom_extendable"] = True
+
+    n.links.loc[ext_dc_links_b, "p_nom"] = p_nom_orig.loc[ext_dc_links_b]
+    n.links.loc[ext_dc_links_b, "p_nom_extendable"] = True
+
+    ## add costs of additional infrastructure to objective value of last iteration
+    obj_links = (
+        n.links[ext_dc_links_b].eval("capital_cost * (p_nom_opt - p_nom_min)").sum()
+    )
+    obj_lines = n.lines.eval("capital_cost * (s_nom_opt - s_nom_min)").sum()
+    n.objective += obj_links + obj_lines
+    n.objective_constant -= obj_links + obj_lines
+
+
+def security_constraint_optimization(
```

Comment:
```suggestion
def optimize_security_constrained(
```
active formulation maybe better?